### DISCUSSION TITLE
using embed_migrations with build environment variable

### DISCUSSION THREAD

**Discussion by WilliamMrc** [diesel-rs/diesel#3295] [Q&A] [answered]
Hello,
I would like to use `embed_migrations` by giving it a migration folder I have dynamically generated.

I Have a build script (build.rs) that generate the migration folder in its `OUT_DIR` and set the `MIGRATIONS_DIR` environment variable. I can get the variable with env! macro and I would like to do something like this : 

`embed_migrations!(env!("MIGRATIONS_DIR"));`

With this code, I get the following error :
```
error: proc-macro derive panicked
  --> src/db.rs:16:1
   |
16 | embed_migrations!(env!("MIGRATIONS_DIR"));
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: message: This is a bug. Please open a Github issue with your invocation of `embed_migrations!
   = note: this error originates in the macro `embed_migrations` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I am note sure this is a bug as the message says. I would like to know if there is a right way of doing this or if I am trying to do something impossible.Thanks !

**Comment by weiznich** [comment, marked as answer]
This sounds like something that's just not supported. Can you provide some additional information about your diesel + diesel_migrations version, as it is currently not clear if you use 1.4.x or any of the 2.0.0 RC versions. Ideally you can provide some minimal example that allows easily to reproduce the issue. 

Otherwise I might be open to merge a fix for this, if it turns out to be broken on 2.0/master as well.

**Reply by WilliamMrc** [reply]
my diesel version is 1.4.8 and diesel_migrarions 1.4.0.

Here is the minimal example : https://github.com/WilliamMrc/diesel_embed_env_minimal_reprodution

After more investigation, I understand that `embed_migrations!(env!("MIGRATIONS_DIR"));` can't work as the env! macro return a `&'static str`. I don't have any suggestion on what could be done to solve this issue  (I am a little bit lost with macro syntax)

To avoid this issue, I just generated my migration folder in `target/migrations` and simply do a `embed_migrations!("target/migrations")`. I just need to delete and regenerate the migration on each build to avoid conflict with a previous build.

**Reply by weiznich** [reply]
You might want to update to diesel 2.0, as this includes a rewrite of the migration framework. That would likely allow you to simply provide your own type implementing  [`MigrationSource`](https://docs.diesel.rs/2.0.x/diesel/migration/trait.MigrationSource.html#tymethod.migrations) and skip any generation step at all.

**Reply by WilliamMrc** [reply]
Thanks, I have updated to diesel 2.0. I am working on the `MigrationSource` implementation.
